pkg/pipeline/source/sdk/appwriter.go

```diff
@@ -354,38 +385,46 @@ func (w *AppWriter) pushBlankFrame(timestamp uint32) error {
 
 func (w *AppWriter) push(packets []*rtp.Packet, blankFrame bool) error {
 	for _, pkt := range packets {
-		// record timestamp diff
-		if w.rtpStep == 0 && !blankFrame && w.lastRTP != 0 && pkt.SequenceNumber == w.lastSN+1 {
-			w.rtpStep = pkt.Timestamp - w.lastRTP
-		}
-
-		// record SN and TS
-		w.lastSN = pkt.SequenceNumber
-		if w.lastRTP > wrapCheck && pkt.Timestamp < wrapCheck {
-			w.rtpWrap++
-		}
-		w.lastRTP = pkt.Timestamp
-
 		if !blankFrame {
 			// update sequence number
 			pkt.SequenceNumber += w.snOffset
+
+			// record max frame size
+			if w.lastTS != 0 && pkt.SequenceNumber == w.lastSN+1 {
+				if frameSize := pkt.Timestamp - w.lastTS; frameSize > w.frameSize && frameSize < w.clockRate/60 {
+					w.frameSize = frameSize
+				}
+			}
+
 			w.translatePacket(pkt)
 		}
 
-		// will return io.EOF if EOS has been sent
+		// reset offsets if needed
+		if w.lastTS != 0 && pkt.SequenceNumber-w.lastSN > maxDropout && w.lastSN-pkt.SequenceNumber > maxDropout {
+			w.logger.Debugw("large SN gap", "previous", w.lastSN, "current", pkt.SequenceNumber)
+			w.resetOffsets(pkt)
+		}
+
+		// get PTS
 		pts, err := w.getPTS(pkt.Timestamp)
 		if err != nil {
 			return err
 		}
 
+		// record timing info
+		w.lastSN = pkt.SequenceNumber
+		w.lastTS = pkt.Timestamp
```

Comment:
I remember some place that ordering happens elsewhere and this is the newest values. Not sure if I am imagining things. Are these supposed to be the highest/latest values? If so, is there a need to check for out-of-order before updating these?

Reply:
yeah, the sample builder acts as a jitter buffer, and this happens after the packets have gone through that